### PR TITLE
DD12 - Slashed dYdX

### PR DESCRIPTION
**Pull request by RCantu92**
# Slashed dYdX Bot

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Bot
- [ ] Fix a bot error
- [ ] Update a bot logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Bot PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the bot properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The bot README is updated
- [x] The bot README sufficiently describes what the bot does
- [x] All the alerts that the bot emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Bot contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks are listed in the `package.json` are the same
- [x] The code does what the README describes
- [ ] Network calls are properly being cached (if duplicated calls can occur)
- [ ] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the bot is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The bot follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [x] Bot can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [x] `repository` is added in the `package.json`

## Comments (optional)

If there is some useful information or important note for reviewing/understanding this bot please describe it here.

### CODE REVIEW

File: DYDX-Bots/Slashed_dYdX/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+This bot detects slashing events in the Safety Module contract.
+
+## Supported Chains
+
+- Ethereum
+
+## Alerts
+
+Describe each of the type of alerts fired by this agent
```
Comment (yasmine9) [inline DYDX-Bots/Slashed_dYdX/README.md:13]:
```suggestion


```

File: DYDX-Bots/Slashed_dYdX/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+- DYDX-12
+  - Fired when `Slashed` event is emitted.
```
Comment (yasmine9) [inline DYDX-Bots/Slashed_dYdX/README.md:16]:
```suggestion
  - Fired when `Slashed` event is emitted in dYdX Safety Module.
```
Small suggestion to complete the description

File: DYDX-Bots/Slashed_dYdX/src/agent.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+export function provideHandleTransaction(networkManager: NetworkData): HandleTransaction {
+  return async (txEvent: TransactionEvent): Promise<Finding[]> => {
+    const findings: Finding[] = [];
+
+    // If the Slashed event is emitted by
+    // the Safety Module contract, create a finding
+    txEvent.filterLog(SLASHED_EVENT, networkManager.safetyModule).map((log) => {
+      findings.push(createFinding(log));
+    });
```
Comment (yasmine9) [inline DYDX-Bots/Slashed_dYdX/src/agent.ts:23]: `map` function is not needed here, we can use `forEach` instead.

File: DYDX-Bots/Slashed_dYdX/src/agent.spec.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+  it("should return 0 findings in empty transactions", async () => {
+    const txEvent: TransactionEvent = new TestTransactionEvent();
+
+    const findings = await handleTransaction(txEvent);
+    expect(findings).toStrictEqual([]);
+  });
+
+  it("should detect a Slashed event emission from both the safety module", async () => {
+    const SlashedLog = MODULE_IFACE.encodeEventLog(MODULE_IFACE.getEvent("Slashed"), [
```
Comment (yasmine9) [inline DYDX-Bots/Slashed_dYdX/src/agent.spec.ts:38]:
```suggestion

  it("should detect a Slashed event emission from the safety module", async () => {
    const SlashedLog = MODULE_IFACE.encodeEventLog(MODULE_IFACE.getEvent("Slashed"), [
```
small typo

File: DYDX-Bots/Slashed_dYdX/src/findings.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { LogDescription, Finding, FindingSeverity, FindingType } from "forta-agent";
+
+export const createFinding = (log: LogDescription) => {
+  return Finding.fromObject({
+    name: "Slash event has occured on dYdX Security Module.",
+    description: "Slashed event was emitted",
```
Comment (yasmine9) [inline DYDX-Bots/Slashed_dYdX/src/findings.ts:6]:
```suggestion
  return Finding.fromObject({
    name: "Slash event has occured on dYdX Safety Module.",
    description: "Slashed event was emitted",
```
Same change is required in the tests file.

File: DYDX-Bots/Slashed_dYdX/src/agent.spec.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+      .setTo(wrongContract)
+      .setFrom(testSender)
+      .addAnonymousEventLog(wrongContract, SlashedLog.data, ...SlashedLog.topics);
+
+    const findings = await handleTransaction(txEvent);
+
+    expect(findings).toStrictEqual([]);
+  });
+
+  it("should not detect another event emission from either the safety nor liquidity modules", async () => {
```
Comment (yasmine9) [inline DYDX-Bots/Slashed_dYdX/src/agent.spec.ts:88]:
```suggestion
  it("should not detect another event emission from the safety module", async () => {
```

**Review by yasmine9 — COMMENTED** [review]
Mainly chore suggestions, in addition to using `forEach` instead of `map` in the bot logic. 
Great Job 👍

**Review by yasmine9 — APPROVED** [review]

File: DYDX-Bots/Slashed_dYdX/README.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+- DYDX-12
+  - Fired when `Slashed` event is emitted in dYdX Safety Module.
+  - Severity is always set to "Info".
+  - Type is always set to "Info".
+  - Metadata includes:
+    - `amount`: Slash amount.
+    - `recipient`: The address to receive the slashed tokens.
+    - `newExchangeRate`: Exchange rate after slashing event.
+  - Addresses is the address from which the event was emitted.
```
Comment (e1Ru1o) [inline DYDX-Bots/Slashed_dYdX/README.md:21]:
```suggestion
  - Addresses contains the address from which the event was emitted.
```

File: DYDX-Bots/Slashed_dYdX/src/agent.spec.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
+import { FindingType, FindingSeverity, Finding, HandleTransaction, ethers, TransactionEvent } from "forta-agent";
```
Comment (e1Ru1o) [inline DYDX-Bots/Slashed_dYdX/src/agent.spec.ts:1]: what about adding a test with multiple findings?

**Pull request merged**